crates/corro-agent/src/agent/handlers.rs

```diff
@@ -401,45 +376,127 @@ pub async fn handle_changes(
     mut rx_changes: CorroReceiver<(ChangeV1, ChangeSource)>,
     mut tripwire: Tripwire,
 ) {
+    const MIN_CHANGES_CHUNK: usize = 2000;
+    let mut queue: VecDeque<(ChangeV1, ChangeSource)> = VecDeque::new();
     let mut buf = vec![];
     let mut count = 0;
 
+    const MAX_CONCURRENT: usize = 3;
+    let mut join_set = JoinSet::new();
+
     let mut max_wait = tokio::time::interval(Duration::from_millis(500));
 
     loop {
+        while count >= MIN_CHANGES_CHUNK && join_set.len() < MAX_CONCURRENT {
+            let mut tmp_count = 0;
+            while let Some((change, src)) = queue.pop_front() {
+                tmp_count += change.len();
+                buf.push((change, src));
+                if tmp_count >= MIN_CHANGES_CHUNK {
+                    break;
+                }
+            }
+
+            if buf.is_empty() {
+                break;
+            }
+
+            debug!(count = %tmp_count, "spawning processing multiple changes from beginning of func");
+            join_set.spawn(util::process_multiple_changes(
+                agent.clone(),
+                bookie.clone(),
+                std::mem::take(&mut buf),
+            ));
+
+            count -= tmp_count;
+        }
+
         tokio::select! {
+            biased;
+
+            // process these first, we don't care about the result,
+            // but we need to drain it to free up concurrency
+            res = join_set.join_next(), if !join_set.is_empty() => {
+                debug!("processed multiple changes concurrently");
+                if let Some(Err(e)) = res {
+                    error!("could not process multiple changes: {e}");
+                }
+                continue;
+            },
```

Comment:
This whole section makes me wonder if we would be better off using a mpmc channel (like `crossbeam_channel`, although not sure how they play with async, otherwise there's also `async-channel` which is mpmc) here instead to concurrently read from the channel and apply changes.

Is batching that important for performance?

This is a tangent but it's frustrating that only the tokio channels implement `recv_many`. I'm wondering if there's some fundamental limitation in mpmc channel designs that don't allow receiving many messages

Reply:
Batching inserts is important for SQLite performance yes. Beginning a transaction, inserting a single row, committing a transaction is a bit expensive. There's also some overhead is sending changes to subscriptions which each have to check if the changes might match their subscription and then apply logic.

How do you see a mpmc channel working here?

Reply:
Ah right, that makes sense.

With an mpmc channel I would invert the batching logic: push changes to a channel, have three tasks pull them from the channel and into a buffer, then apply in-line when there's enough changes. If memory drop pressure is an issue we could even make them dedicated threads (although that might be overkill).

It _should_ result in a similar performance profile but simplify the big tangled function which receives from the channel, then applies to a buffer, then spawn tasks with buffers as needed.
(Changes wouldn't be applied in order, but I assume the order that changes are appearing on the node isn't in-order anyway so that wouldn't matter much?)